File: momentra/momentra/Shell/PersonalEmpty/RelationshipsActivityModels.swift
import Foundation
import SwiftUI

struct RelationshipsActivityItem: Identifiable, Equatable {
    var id: String
    var title: String
    var whenLabel: String
    var impact: String
    var emoji: String
    var relationship: String
    var notes: String
    var tags: [String]
    var filter: String
}

enum RelationshipsActivityModels {
    /// Empty API / failure → empty list. Never invent demo rows (S2 G2).
    static func from(api items: [APIClient.ActivityItemPayload]) -> [RelationshipsActivityItem] {
        items.map { dto in
            let meta = visual(for: dto.activityCode, title: dto.title)
            return RelationshipsActivityItem(
                id: dto.occurredAt + dto.title + (dto.activityPayload?.activityId ?? ""),
                title: dto.title.isEmpty ? meta.title : dto.title,
                whenLabel: formatOccurred(dto.occurredAt),
                impact: "",
                emoji: meta.emoji,
                relationship: meta.filter,
                notes: "",
                tags: [],
                filter: meta.filter
            )
        }
    }

    private static func visual(for code: String, title: String) -> (title: String, emoji: String, filter: String) {
        let c = code.uppercased()
        if c.contains("CONNECT") || title.localizedCaseInsensitiveContains("connection") {
            return ("Connection", "💬", "Partner")
        }
        if c.contains("SUPPORT") || title.localizedCaseInsensitiveContains("support") {
            return ("Support", "🫶", "Partner")
        }
        if c.contains("SHARED") {
            return ("Shared experience", "🤝", "Partner")
        }
        if c.contains("INVEST") {
            return ("Investment", "📈", "Partner")
        }
        if c.contains("FAMILY") || title.localizedCaseInsensitiveContains("family") {
            return ("Family", "📞", "Family")
        }
        if c.contains("FRIEND") {
            return ("Friends", "🤝", "Friends")
        }
        if c.contains("SELF") || c.contains("REFLECT") || c.contains("INTERACTION") {
            return ("Check-in", "🪞", "Self")
        }
        return ("Activity", "❤️", "Partner")
    }

    private static func formatOccurred(_ iso: String) -> String {
        let f = ISO8601DateFormatter()
        f.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        let date = f.date(from: iso) ?? ISO8601DateFormatter().date(from: iso)
        guard let date else { return iso }
        let out = DateFormatter()
        out.dateFormat = "MMM d, h:mm a"
        return out.string(from: date)
    }
}

File: apk/app/src/main/java/com/example/momentra/ui/shell/personal/relationships/create/RelationshipsActivityModels.kt
package com.example.momentra.ui.shell.personal.relationships.create

import com.example.momentra.data.api.ActivityItemDto
import java.util.UUID

/** Relationships activity row — Figma `1036:7697` / Pulse preview. */
data class RelationshipsActivityItem(
    val id: String = UUID.randomUUID().toString(),
    val title: String,
    val whenLabel: String,
    val impact: String,
    val emoji: String,
    val relationship: String = "Partner",
    val notes: String = "",
    val tags: List<String> = emptyList(),
    val filter: String = "Partner",
)

fun mapActivityDtosToRelationships(items: List<ActivityItemDto>): List<RelationshipsActivityItem> {
    if (items.isEmpty()) return emptyList()
    return items.map { dto ->
        val meta = relationshipsActivityVisual(dto.activityCode, dto.title)
        RelationshipsActivityItem(
            id = dto.occurredAt + dto.title,
            title = dto.title.ifBlank { meta.title },
            whenLabel = formatRelationshipsOccurredAt(dto.occurredAt),
            impact = meta.impact,
            emoji = meta.emoji,
            relationship = meta.filter,
            filter = meta.filter,
        )
    }
}

private data class RelVisual(val title: String, val emoji: String, val impact: String, val filter: String)

private fun relationshipsActivityVisual(code: String, title: String): RelVisual {
    val c = code.uppercase()
    return when {
        c.contains("CONNECT") || title.contains("connection", true) ->
            RelVisual(title.ifBlank { "Connection" }, "💬", "Connection", "Partner")
        c.contains("SUPPORT") || title.contains("support", true) ->
            RelVisual(title.ifBlank { "Support" }, "🫶", "Support", "Partner")
        c.contains("SHARED") || title.contains("shared", true) ->
            RelVisual(title.ifBlank { "Shared experience" }, "🤝", "Shared", "Partner")
        c.contains("INVEST") ->
            RelVisual(title.ifBlank { "Investment" }, "💝", "Investment", "Partner")
        c.contains("FAMILY") || title.contains("family", true) ->
            RelVisual(title.ifBlank { "Family" }, "📞", "Family", "Family")
        c.contains("FRIEND") ->
            RelVisual(title.ifBlank { "Friend time" }, "🤝", "Friends", "Friends")
        c.contains("SELF") || c.contains("REFLECT") || c.contains("INTERACTION") ->
            RelVisual(title.ifBlank { "Interaction" }, "🪞", "Interaction", "Self")
        else -> RelVisual(title.ifBlank { "Activity" }, "❤️", "Logged", "Partner")
    }
}

private fun formatRelationshipsOccurredAt(iso: String): String = try {
    val instant = java.time.Instant.parse(iso)
    java.time.format.DateTimeFormatter.ofPattern("MMM d, h:mm a", java.util.Locale.getDefault())
        .withZone(java.time.ZoneId.systemDefault())
        .format(instant)
} catch (_: Exception) {
    iso
}
